CONFIG.md
# Configuration

## Overview

The configuration file for CosmoGuard allows fine-grained control over various aspects,
including host settings, ports, caching, metrics, and specific rules for LCD, gRPC, RPC, JSON-RPC, and EVM.
Most fields are optional and default to specified values.

## Configuration Options

| Option                                | Default Value     | Description                                                                                                         |
|---------------------------------------|-------------------|---------------------------------------------------------------------------------------------------------------------|
| **Global Settings**                   |                   |                                                                                                                     |
| `host`                                | `0.0.0.0`         | The host address the CosmoGuard listens on.                                                                         |
| `rpcPort`                             | `16657`           | Port for Tendermint RPC.                                                                                            |
| `lcdPort`                             | `11317`           | Port for Cosmos REST API (LCD).                                                                                     |
| `grpcPort`                            | `19090`           | Port for gRPC.                                                                                                      |
| `enableEvm`                           | `false`           | Enable or disable EVM support.                                                                                      |
| `evmRpcPort`                          | `18545`           | Port for EVM JSON-RPC.                                                                                              |
| `evmRpcWsPort`                        | `18546`           | Port for EVM WebSocket.                                                                                             |
| **Node Settings**                     |                   |                                                                                                                     |
| `node.host`                           | `127.0.0.1`       | The node's host address.                                                                                            |
| `node.rpcPort`                        | `26657`           | Port for the node's Tendermint RPC.                                                                                 |
| `node.lcdPort`                        | `1317`            | Port for the node's LCD.                                                                                            |
| `node.grpcPort`                       | `9090`            | Port for the node's gRPC.                                                                                           |
| `node.evmRpcPort`                     | `8545`            | Port for the node's EVM JSON-RPC.                                                                                   |
| `node.evmRpcWsPort`                   | `8546`            | Port for the node's EVM WebSocket.                                                                                  |
| **Caching Settings**                  |                   |                                                                                                                     |
| `cache.ttl`                           | `5s`              | Global time-to-live for cached responses.                                                                           |
| `cache.redis`                         | (In-memory cache) | Connection URL for Redis if used for caching.                                                                       |
| `cache.redis-sentinel.master_name`    |                   | The master name for Redis Sentinel.                                                                                 |
| `cache.redis-sentinel.sentinel_addrs` |                   | A list of sentinel addresses for Redis Sentinel.                                                                    |
| `cache.key`                           |                   | Optional key prefix for cache. Useful for splitting different caches in same database. Not used on in-memory cache. |
| **Metrics Settings**                  |                   |                                                                                                                     |
| `metrics.enable`                      | `true`            | Enable or disable metrics collection.                                                                               |
| `metrics.port`                        | `9001`            | Port for metrics.                                                                                                   |
| **LCD (REST API) Settings**           |                   |                                                                                                                     |
| `lcd.default`                         | `deny`            | Default action for LCD requests.                                                                                    |
| `lcd.rules`                           | (Empty)           | List of [HTTP Rules](#http-rule) for handling LCD requests.                                                         |
| **gRPC Settings**                     |                   |                                                                                                                     |
| `grpc.default`                        | `deny`            | Default action for gRPC requests.                                                                                   |
| `grpc.rules`                          | (Empty)           | List of [gRPC Rules](#grpc-rule) for handling gRPC requests.                                                        |
| **RPC Settings**                      |                   |                                                                                                                     |
| `rpc.default`                         | `deny`            | Default action for RPC requests.                                                                                    |
| `rpc.rules`                           | (Empty)           | List of [HTTP Rules](#http-rule) for handling RPC requests.                                                         |
| **JSON-RPC Settings**                 |                   |                                                                                                                     |
| `rpc.jsonrpc.default`                 | `deny`            | Default action for JSON-RPC requests.                                                                               |
| `rpc.jsonrpc.webSocketEnabled`        | `true`            | Enable or disable WebSocket support.                                                                                |
| `rpc.jsonrpc.webSocketConnections`    | `10`              | Maximum number of WebSocket connections.                                                                            |
| `rpc.jsonrpc.rules`                   | (Empty)           | List of [JSON-RPC Rules](#json-rpc-rule) for handling JSON-RPC requests.                                            |
| **EVM Settings**                      |                   |                                                                                                                     |
| `evm.rpc.default`                     | `deny`            | Default action for EVM RPC requests.                                                                                |
| `evm.rpc.rules`                       | (Empty)           | List of [JSON-RPC Rules](#json-rpc-rule) for handling EVM RPC requests.                                             |
| `evm.ws.default`                      | `deny`            | Default action for EVM WebSocket requests.                                                                          |
| `evm.ws.webSocketConnections`         | `10`              | Maximum number of WebSocket connections.                                                                            |
| `evm.ws.rules`                        | (Empty)           | List of [JSON-RPC Rules](#json-rpc-rule) for handling EVM WebSocket requests.                                       |

## Rules

### HTTP rule

| Option         | Default Value | Description                                                                                                          |
|----------------|---------------|----------------------------------------------------------------------------------------------------------------------|
| `priority`     | `1000`        | The priority of the rule. Lower numbers indicate higher priority.                                                    |
| `action`       | -             | The action to be taken for matching requests: either `allow` or `deny`.                                              |
| `paths`        | (Empty)       | List of paths this rule applies to. An empty list means all paths are included. **Wildcards supported**.             |
| `methods`      | (Empty)       | List of HTTP methods (e.g., `GET`, `POST`) this rule applies to. An empty list means all methods are included.       |
| `cache`        |               | Caching settings for the rule.                                                                                       |
| `cache.enable` | `false`       | Enable or disable caching for this rule.                                                                             |
| `cache.ttl`    | `5s`          | Time-to-live for the cache, which can override the global cache TTL. Default is the global setting if not specified. |

#### Example

```yaml
rules:
  - priority: 9999
    action: deny
    paths: [ ]
    methods: [ ]

  - priority: 1000
    action: allow
    paths:
      - /status
      - /block/**
    methods:
      - GET
    cache:
      enable: true
      ttl: 20s
```

### JSON-RPC rule

| Option         | Default Value | Description                                                                                                                                                                  |
|----------------|---------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `priority`     | `1000`        | The priority of the rule. Lower numbers indicate higher priority.                                                                                                            |
| `action`       | `allow`       | The action to be taken for matching requests: either `allow` or `deny`.                                                                                                      |
| `methods`      | (Empty)       | List of JSON-RPC methods this rule applies to. An empty list means all methods are included. **Wildcards supported**.                                                        |
| `params`       | (Empty)       | Dictionary of parameters to match specific JSON-RPC requests. Empty or missing means all parameters are included. **Wildcards supported (on params values only, not keys)**. |
| `cache`        |               | Caching settings for the rule.                                                                                                                                               |
| `cache.enable` | `false`       | Enable or disable caching for this rule.                                                                                                                                     |
| `cache.ttl`    | `5s`          | Time-to-live for the cache, which can override the global cache TTL. Default is the global setting if not specified.                                                         |

#### Example

```yaml
rules:
  - priority: 9999
    action: deny
    methods: [ ]
    params: { }

  - action: allow
    methods: [ "subscribe", "unsubscribe", "unsubscribe_all" ]

  - action: allow
    methods: [ "status" ]
    cache:
      enable: true

  - action: allow
    methods: [ "abci_query" ]
    params:
      path: /cosmos.bank.v1beta1.Query/AllBalances
    cache:
      enable: true
      ttl: 2s

  - action: allow
    methods: [ "abci_query" ]
    params:
      path: /cosmos.distribution.v1beta1.Query/**
```

### gRPC rule

| Option     | Default Value | Description                                                                                                       |
|------------|---------------|-------------------------------------------------------------------------------------------------------------------|
| `priority` | `1000`        | The priority of the rule. Lower numbers indicate higher priority.                                                 |
| `action`   | `allow`       | The action to be taken for matching requests: either `allow` or `deny`.                                           |
| `methods`  | (Empty)       | List of gRPC methods this rule applies to. An empty list means all methods are included. **Wildcards supported**. |

#### Example

```yaml
rules:
  - priority: 9999
    action: deny
    methods: [ ]

  - action: allow
    methods:
      - /cosmos.bank.v1beta1.Query/AllBalances

  - action: allow
    methods:
      - /cosmos.staking.v1beta1.Query/**
```

pkg/cosmoguard/config.go
package cosmoguard

import (
	"fmt"
	"os"
	"sort"
	"time"

	"github.com/creasty/defaults"
	"gopkg.in/yaml.v3"
)

const (
	cacheKey  = "Cache"
	cacheHit  = "hit"
	cacheMiss = "miss"
)

type Config struct {
	Host         string            `yaml:"host,omitempty" default:"0.0.0.0"`
	RpcPort      int               `yaml:"rpcPort,omitempty" default:"16657"`
	LcdPort      int               `yaml:"lcdPort,omitempty" default:"11317"`
	GrpcPort     int               `yaml:"grpcPort,omitempty" default:"19090"`
	EnableEvm    bool              `yaml:"enableEvm,omitempty" default:"false"`
	EvmRpcPort   int               `yaml:"evmRpcPort,omitempty" default:"18545"`
	EvmRpcWsPort int               `yaml:"evmRpcWsPort,omitempty" default:"18546"`
	Node         NodeConfig        `yaml:"node,omitempty"`
	Cache        CacheGlobalConfig `yaml:"cache,omitempty"`
	LCD          LcdConfig         `yaml:"lcd,omitempty"`
	RPC          RpcConfig         `yaml:"rpc,omitempty"`
	GRPC         GrpcConfig        `yaml:"grpc,omitempty"`
	EVM          EvmConfig         `yaml:"evm,omitempty"`
	Metrics      MetricsConfig     `yaml:"metrics,omitempty"`
}

type NodeConfig struct {
	Host         string `yaml:"host,omitempty" default:"127.0.0.1"`
	RpcPort      int    `yaml:"rpcPort,omitempty" default:"26657"`
	LcdPort      int    `yaml:"lcdPort,omitempty" default:"1317"`
	GrpcPort     int    `yaml:"grpcPort,omitempty" default:"9090"`
	EvmRpcPort   int    `yaml:"evmRpcPort,omitempty" default:"8545"`
	EvmRpcWsPort int    `yaml:"evmRpcWsPort,omitempty" default:"8546"`
}

type CacheGlobalConfig struct {
	TTL           time.Duration        `yaml:"ttl,omitempty" default:"5s"`
	Redis         *string              `yaml:"redis,omitempty"`
	RedisSentinel *RedisSentinelConfig `yaml:"redis-sentinel,omitempty"`
	Key           string               `yaml:"key"`
}

type RedisSentinelConfig struct {
	MasterName    string   `yaml:"master_name,omitempty"`
	SentinelAddrs []string `yaml:"sentinel_addrs,omitempty"`
}

type RuleCache struct {
	Enable           bool          `yaml:"enable,omitempty"`
	TTL              time.Duration `yaml:"ttl,omitempty"`
	CacheError       bool          `yaml:"cacheError,omitempty"`
	CacheEmptyResult bool          `yaml:"cacheEmptyResult,omitempty"`
}

type LcdConfig struct {
	Default RuleAction  `yaml:"default,omitempty" default:"deny"`
	Rules   []*HttpRule `yaml:"rules,omitempty"`
}

type RpcConfig struct {
	Default              RuleAction    `yaml:"default,omitempty" default:"deny"`
	Rules                []*HttpRule   `yaml:"rules,omitempty"`
	JsonRpc              JsonRpcConfig `yaml:"jsonrpc,omitempty"`
	WebSocketEnabled     bool          `yaml:"webSocketEnabled,omitempty" default:"true"`
	WebSocketConnections int           `yaml:"webSocketConnections,omitempty" default:"10"`
}

type JsonRpcConfig struct {
	Default RuleAction     `yaml:"default,omitempty" default:"deny"`
	Rules   []*JsonRpcRule `yaml:"rules,omitempty"`
}

type GrpcConfig struct {
	Default RuleAction  `yaml:"default,omitempty" default:"deny"`
	Rules   []*GrpcRule `yaml:"rules,omitempty"`
}

type MetricsConfig struct {
	Enable bool `yaml:"enable" default:"true"`
	Port   int  `yaml:"port,omitempty" default:"9001"`
}

type EvmConfig struct {
	RPC EvmRpcConfig   `yaml:"rpc,omitempty"`
	WS  EvmRpcWsConfig `yaml:"ws,omitempty"`
}

type EvmRpcConfig struct {
	Default   RuleAction     `yaml:"default,omitempty" default:"deny"`
	Rules     []*JsonRpcRule `yaml:"rules,omitempty"`
	HttpRules []*HttpRule    `yaml:"httpRules,omitempty"`
}

type EvmRpcWsConfig struct {
	Default              RuleAction     `yaml:"default,omitempty" default:"deny"`
	Rules                []*JsonRpcRule `yaml:"rules,omitempty"`
	WebSocketConnections int            `yaml:"webSocketConnections,omitempty" default:"10"`
}

func ReadConfigFromFile(path string) (*Config, error) {
	f, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("error reading config file: %v", err)
	}
	var cfg Config
	err = yaml.Unmarshal(f, &cfg)
	if err != nil {
		return nil, fmt.Errorf("error in config file unmarshal: %v", err)
	}

	// Sort rules by priority and compile rules
	if cfg.LCD.Rules != nil {
		sort.Slice(cfg.LCD.Rules, func(i, j int) bool {
			return cfg.LCD.Rules[i].Priority < cfg.LCD.Rules[j].Priority
		})
		for _, rule := range cfg.LCD.Rules {
			rule.Compile()
		}
	}
	if cfg.GRPC.Rules != nil {
		sort.Slice(cfg.GRPC.Rules, func(i, j int) bool {
			return cfg.GRPC.Rules[i].Priority < cfg.GRPC.Rules[j].Priority
		})
		for _, rule := range cfg.GRPC.Rules {
			rule.Compile()
		}
	}

	if cfg.RPC.Rules != nil {
		sort.Slice(cfg.RPC.Rules, func(i, j int) bool {
			return cfg.RPC.Rules[i].Priority < cfg.RPC.Rules[j].Priority
		})
	}
	for _, rule := range cfg.RPC.Rules {
		rule.Compile()
	}

	if cfg.RPC.JsonRpc.Rules != nil {
		sort.Slice(cfg.RPC.JsonRpc.Rules, func(i, j int) bool {
			return cfg.RPC.JsonRpc.Rules[i].Priority < cfg.RPC.JsonRpc.Rules[j].Priority
		})
		for _, rule := range cfg.RPC.JsonRpc.Rules {
			rule.Compile()
		}
	}

	if cfg.EVM.RPC.Rules != nil {
		sort.Slice(cfg.EVM.RPC.Rules, func(i, j int) bool {
			return cfg.EVM.RPC.Rules[i].Priority < cfg.EVM.RPC.Rules[j].Priority
		})
		for _, rule := range cfg.EVM.RPC.Rules {
			rule.Compile()
		}
	}

	if cfg.EVM.RPC.HttpRules != nil {
		sort.Slice(cfg.EVM.RPC.HttpRules, func(i, j int) bool {
			return cfg.EVM.RPC.HttpRules[i].Priority < cfg.EVM.RPC.HttpRules[j].Priority
		})
		for _, rule := range cfg.EVM.RPC.HttpRules {
			rule.Compile()
		}
	}

	if cfg.EVM.WS.Rules != nil {
		sort.Slice(cfg.EVM.WS.Rules, func(i, j int) bool {
			return cfg.EVM.WS.Rules[i].Priority < cfg.EVM.WS.Rules[j].Priority
		})
		for _, rule := range cfg.EVM.WS.Rules {
			rule.Compile()
		}
	}

	return &cfg, defaults.Set(&cfg)
}
